features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/components/MessageShieldView.kt

```diff
@@ -0,0 +1,131 @@
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.graphics.vector.ImageVector
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.unit.dp
+import io.element.android.compound.theme.ElementTheme
+import io.element.android.compound.tokens.generated.CompoundIcons
+import io.element.android.libraries.designsystem.preview.ElementPreview
+import io.element.android.libraries.designsystem.preview.PreviewsDayNight
+import io.element.android.libraries.designsystem.theme.components.Icon
+import io.element.android.libraries.designsystem.theme.components.Text
+import io.element.android.libraries.matrix.api.timeline.item.event.MessageShield
+import io.element.android.libraries.matrix.api.timeline.item.event.isCritical
+import io.element.android.libraries.ui.strings.CommonStrings
+
+@Composable
+internal fun MessageShieldView(
+    shield: MessageShield,
+    modifier: Modifier = Modifier
+) {
+    Row(
+        verticalAlignment = Alignment.CenterVertically,
+        modifier = modifier,
+    ) {
+        Icon(
+            imageVector = shield.toIcon(),
+            contentDescription = null,
+            modifier = Modifier.size(15.dp),
+            tint = shield.toIconColor(),
+        )
+        Spacer(modifier = Modifier.size(4.dp))
+        Text(
+            text = shield.toText(),
+            style = ElementTheme.typography.fontBodySmMedium,
+            color = shield.toTextColor()
+        )
+    }
+}
+
+@Composable
+internal fun MessageShield.toIconColor(): Color {
+    return when (isCritical()) {
+        true -> ElementTheme.colors.iconCriticalPrimary
+        false -> ElementTheme.colors.iconSecondary
+    }
+}
+
+@Composable
+private fun MessageShield.toTextColor(): Color {
+    return when (isCritical()) {
+        true -> ElementTheme.colors.textCriticalPrimary
+        false -> ElementTheme.colors.textSecondary
+    }
+}
+
+@Composable
+internal fun MessageShield.toText(): String {
+    return stringResource(
+        id = when (this) {
+            is MessageShield.AuthenticityNotGuaranteed -> CommonStrings.event_shield_reason_authenticity_not_guaranteed
+            is MessageShield.UnknownDevice -> CommonStrings.event_shield_reason_unknown_device
+            is MessageShield.UnsignedDevice -> CommonStrings.event_shield_reason_unsigned_device
+            is MessageShield.UnverifiedIdentity -> CommonStrings.event_shield_reason_unverified_identity
+            is MessageShield.SentInClear -> CommonStrings.event_shield_reason_sent_in_clear
+        }
+    )
+}
+
+@Composable
+internal fun MessageShield.toIcon(): ImageVector {
+    return when (this) {
+        is MessageShield.AuthenticityNotGuaranteed,
+        is MessageShield.UnverifiedIdentity -> CompoundIcons.Admin()
+        is MessageShield.UnknownDevice,
+        is MessageShield.UnsignedDevice -> CompoundIcons.HelpSolid()
+        is MessageShield.SentInClear -> CompoundIcons.KeyOff()
```

Comment:
TODO: need to check this mapping.

Reply:
This should be the final mapping: https://github.com/element-hq/element-x-ios/blob/cf35056c787fbae4a219e2a442738b6a84164a95/ElementX/Sources/Services/Timeline/TimelineItemContent/EncryptionAuthenticity.swift#L59-L65 (I realise now that I linked you to a bunch of PRs that iteratively changed it over time).

Reply:
Thanks. I have updated the code.
But there is no shield at all, which is against the SDK recommendation [here](https://github.com/matrix-org/matrix-rust-sdk/blob/main/crates/matrix-sdk-common/src/deserialized_responses.rs#L196) and [here](https://github.com/matrix-org/matrix-rust-sdk/blob/main/crates/matrix-sdk-common/src/deserialized_responses.rs#L204).

Now I have this (from the Preview):

<img width="486" alt="image" src="https://github.com/user-attachments/assets/23a08420-8d01-4771-8f50-3233e4822d1b">

@BillCarsonFr can you confirm please?